File: dnn-providers/hip-kernel-provider/rocke/library/tests/parity/conv_implicit_gemm_wgrad_emit.py
#!/usr/bin/env python3
# Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
# SPDX-License-Identifier: MIT
#
# tests/parity/conv_implicit_gemm_wgrad_emit.py -- Python reference emitter
# for the implicit-GEMM backward-weight convolution parity harness.
# Selects one of N sampled spec configs by argv[1], builds the WgradConvSpec,
# builds the kernel via build_implicit_gemm_conv_wgrad(spec, arch=<cfg arch>)
# and prints lower_kernel_to_llvm(arch=<cfg arch>) to stdout so it can be
# byte-compared with the C emitter conv_implicit_gemm_wgrad_emit.c.
#
# Config index map:
#   0  -- 3x3 2-D, mem/default, gfx950
#   1  -- 3x3 2-D, mem/cshuffle, gfx950
#   2  -- split-K=4 fp16 output, gfx950
#   3  -- 1x1 2-D, mem/default, gfx950
#   4  -- 128x128 tile, compv4/default, gfx950
#   5  -- WMMA gfx1151
#   6  -- WMMA gfx1201
#   7  -- split-K=4 fp32 output, gfx950
#   8  -- 3-D conv (Z/Di), mem/default, gfx950
#   9  -- split-K=4 bf16 output (packed bf16 atomic + accumulation-error path), gfx950
#   10 -- chiplet swizzle enabled, gfx950
#   11 -- K-outer LDS + ds_read_b64_tr_b16 transpose reads, gfx950
#   12 -- K-outer + async_dma (direct global->LDS load), gfx950
#   13 -- K-outer with the 16x16x16 atom (4 operand elements per lane), gfx950
#   14 -- pipeline="basic" unrolled global-read/compute overlap, gfx950
#   15 -- split-K=4, two_stage=True (workspace-store epilogue), fp16, gfx950
#   16 -- split-K=4, two_stage=True (workspace-store epilogue), fp16, gfx942
#   17 -- gfx1250 wave32 WMMA 16x16x32 K-outer (ds_load_tr16_b128 transpose reads)
#   (async_dma omitted: C++ async load path does not yet honour the wgrad A-descriptor
#    override, so it would produce different IR and break the byte-identity gate)
#
# Negative cases (configs 100+) verify that invalid specs are rejected:
#   100 -- odd C with fp16 split-K (must raise ValueError)
#   102 -- split_k > 1 on RDNA gfx1151 (must raise ValueError)
#   103 -- two_stage=True with split_k=1 (must raise ValueError)
# (These illustrate the validator contract. The C emitter defines only cases
# 0-17, so run_diff.py stops at the shared END before reaching 100+; these
# configs are not exercised by the differential gate.)
from kernels.common.conv_implicit_gemm_wgrad import (
    WgradConvSpec,
    build_implicit_gemm_conv_wgrad,
)
from kernels.common._conv_implicit_gemm_common import ConvProblem
from _emit_common import run_emit


def _spec(idx: int):
    """Return (spec, arch) for config index `idx`.

    For negative cases (idx >= 100) the spec is expected to raise ValueError
    when built (run this file with that index to confirm it raises). These are
    illustrative of the validator contract; no differential harness consumes
    them (see the module header).
    """
    if idx == 0:
        # Baseline: 3x3 conv, mem pipeline, default epilogue, gfx950.
        p = ConvProblem(N=8, Hi=56, Wi=56, C=64, K=64, Y=3, X=3)
        return (
            WgradConvSpec(
                problem=p,
                tile_m=64,
                tile_n=64,
                tile_k=64,
                warp_m=2,
                warp_n=2,
                warp_tile_m=32,
                warp_tile_n=32,
                warp_tile_k=16,
                pipeline="mem",
                epilogue="default",
            ),
            "gfx950",
        )
    if idx == 1:
        # cshuffle epilogue, gfx950.
        p = ConvProblem(N=8, Hi=56, Wi=56, C=64, K=64, Y=3, X=3)
        return (
            WgradConvSpec(
                problem=p,
                tile_m=64,
                tile_n=64,
                tile_k=64,
                warp_m=2,
                warp_n=2,
                warp_tile_m=32,
                warp_tile_n=32,
                warp_tile_k=16,
                pipeline="mem",
                epilogue="cshuffle",
            ),
            "gfx950",
        )
    if idx == 2:
        # Split-K=4 with fp16 output, gfx950.
        p = ConvProblem(N=8, Hi=56, Wi=56, C=64, K=64, Y=3, X=3)
        return (
            WgradConvSpec(
                problem=p,
                tile_m=64,
                tile_n=64,
                tile_k=64,
                warp_m=2,
                warp_n=2,
                warp_tile_m=32,
                warp_tile_n=32,
                warp_tile_k=16,
                pipeline="mem",
                epilogue="default",
                split_k=4,
            ),
            "gfx950",
        )
    if idx == 3:
        # 1x1 conv, mem pipeline, gfx950.
        p = ConvProblem(N=8, Hi=56, Wi=56, C=64, K=64, Y=1, X=1)
        return (
            WgradConvSpec(
                problem=p,
                tile_m=64,
                tile_n=64,
                tile_k=64,
                warp_m=2,
                warp_n=2,
                warp_tile_m=32,
                warp_tile_n=32,
                warp_tile_k=16,
                pipeline="mem",
                epilogue="default",
            ),
            "gfx950",
        )
    if idx == 4:
        # Larger tile, compv4 pipeline, gfx950.
        p = ConvProblem(N=8, Hi=56, Wi=56, C=64, K=64, Y=3, X=3)
        return (
            WgradConvSpec(
                problem=p,
                tile_m=128,
                tile_n=128,
                tile_k=64,
                warp_m=2,
                warp_n=2,
                warp_tile_m=32,
                warp_tile_n=32,
                warp_tile_k=16,
                pipeline="compv4",
                epilogue="default",
            ),
            "gfx950",
        )
    if idx in (5, 6):
        # WMMA wave32 RDNA targets: 16x16x16 / mem / default.
        arch = {5: "gfx1151", 6: "gfx1201"}[idx]
        p = ConvProblem(N=8, Hi=56, Wi=56, C=64, K=64, Y=3, X=3)
        return (
            WgradConvSpec(
                problem=p,
                tile_m=64,
                tile_n=64,
                tile_k=64,
                warp_m=2,
                warp_n=2,
                warp_tile_m=16,
                warp_tile_n=16,
                warp_tile_k=16,
                wave_size=32,
                pipeline="mem",
                epilogue="default",
            ),
            arch,
        )
    if idx == 7:
        # Split-K=4 with fp32 output, gfx950.
        p = ConvProblem(N=8, Hi=56, Wi=56, C=64, K=64, Y=3, X=3)
        from kernels.common._conv_implicit_gemm_common import ConvDataSpec

        return (
            WgradConvSpec(
                problem=p,
                data=ConvDataSpec(dtype_a="fp16", dtype_b="fp16", dtype_d="fp32"),
                tile_m=64,
                tile_n=64,
                tile_k=64,
                warp_m=2,
                warp_n=2,
                warp_tile_m=32,
                warp_tile_n=32,
                warp_tile_k=16,
                pipeline="mem",
                epilogue="default",
                split_k=4,
            ),
            "gfx950",
        )
    if idx == 8:
        # 3-D convolution (Z/Di path), mem/default, gfx950.
        p = ConvProblem(
            N=4,
            Hi=14,
            Wi=14,
            Di=14,
            C=32,
            K=32,
            Y=3,
            X=3,
            Z=3,
            sH=1,
            sW=1,
            sD=1,
            pH=1,
            pW=1,
            pD=1,
            dH=1,
            dW=1,
            dD=1,
        )
        return (
            WgradConvSpec(
                problem=p,
                tile_m=64,
                tile_n=64,
                tile_k=64,
                warp_m=2,
                warp_n=2,
                warp_tile_m=32,
                warp_tile_n=32,
                warp_tile_k=16,
                pipeline="mem",
                epilogue="default",
            ),
            "gfx950",
        )
    if idx == 9:
        # Split-K=4 with bf16 output -- exercises packed bf16 atomic +
        # accumulation-error path (the newest codegen, previously untested).
        p = ConvProblem(N=8, Hi=56, Wi=56, C=64, K=64, Y=3, X=3)
        from kernels.common._conv_implicit_gemm_common import ConvDataSpec

        return (
            WgradConvSpec(
                problem=p,
                data=ConvDataSpec(dtype_a="fp16", dtype_b="fp16", dtype_d="bf16"),
                tile_m=64,
                tile_n=64,
                tile_k=64,
                warp_m=2,
                warp_n=2,
                warp_tile_m=32,
                warp_tile_n=32,
                warp_tile_k=16,
                pipeline="mem",
                epilogue="default",
                split_k=4,
            ),
            "gfx950",
        )
    if idx == 10:
        # Chiplet swizzle enabled, gfx950.
        p = ConvProblem(N=8, Hi=56, Wi=56, C=64, K=64, Y=3, X=3)
        return (
            WgradConvSpec(
                problem=p,
                tile_m=64,
                tile_n=64,
                tile_k=64,
                warp_m=2,
                warp_n=2,
                warp_tile_m=32,
                warp_tile_n=32,
                warp_tile_k=16,
                pipeline="mem",
                epilogue="default",
                chiplet_swizzle=True,
            ),
            "gfx950",
        )

    if idx == 11:
        # K-outer LDS tile + gfx950 ds_read_b64_tr_b16 transpose reads.
        # Exercises the swapped loader axes, the K-outer smem shapes and the
        # transpose-read fragment feed in one config.
        p = ConvProblem(N=8, Hi=56, Wi=56, C=64, K=64, Y=3, X=3)
        return (
            WgradConvSpec(
                problem=p,
                tile_m=64,
                tile_n=64,
                tile_k=64,
                warp_m=2,
                warp_n=2,
                warp_tile_m=32,
                warp_tile_n=32,
                warp_tile_k=16,
                pipeline="mem",
                # cshuffle, not default: the validator rejects dtype_d in
                # (fp16, bf16) with epilogue="default", so a "default" config
                # would land as BOTH_REJECTED and the gate would never actually
                # compare the two engines.
                epilogue="cshuffle",
                lds_k_outer=True,
            ),
            "gfx950",
        )

    if idx == 12:
        from kernels.common._conv_implicit_gemm_common import ConvDataSpec

        # K-outer + direct load. Exercises the async loader on the swapped tile
        # axes, the contig_cols guard and the packed (pad-0) LDS shape.
        p = ConvProblem(N=8, Hi=56, Wi=56, C=64, K=64, Y=3, X=3)
        return (
            WgradConvSpec(
                problem=p,
                tile_m=64,
                tile_n=64,
                tile_k=64,
                warp_m=2,
                warp_n=2,
                warp_tile_m=32,
                warp_tile_n=32,
                warp_tile_k=16,
                pipeline="mem",
                # fp32 output + the direct-store epilogue. split-K with the
                # cshuffle atomic epilogue is separately divergent between the
                # engines (reproducible with neither async nor K-outer), so this
                # config isolates the async / K-outer path under split-K.
                epilogue="default",
                data=ConvDataSpec(dtype_d="fp32"),
                lds_k_outer=True,
                async_dma=True,
                split_k=4,
            ),
            "gfx950",
        )

    if idx == 13:
        # K-outer with the 16x16x16 atom: the only 16-bit atom whose MFMA
        # operand is 4 elements per lane rather than 8, so the transpose-read
        # k-stride between lane groups is 4. Configs 11 and 12 both use
        # 32x32x16 (8 per lane) and cannot catch a stride hardcoded to 8.
        p = ConvProblem(N=8, Hi=56, Wi=56, C=64, K=64, Y=3, X=3)
        return (
            WgradConvSpec(
                problem=p,
                tile_m=64,
                tile_n=64,
                tile_k=16,
                warp_m=2,
                warp_n=2,
                warp_tile_m=16,
                warp_tile_n=16,
                warp_tile_k=16,
                pipeline="mem",
                epilogue="cshuffle",
                lds_k_outer=True,
            ),
            "gfx950",
        )

    if idx == 14:
        # pipeline="basic": the CK pipeline_basic loop, unrolled at build time
        # with the global read for tile it+1 issued before the MFMA for tile it
        # and the LDS write deferred past the second barrier. Needs a
        # compile-time trip count, hence the fixed split_k.
        from kernels.common._conv_implicit_gemm_common import ConvDataSpec

        p = ConvProblem(N=8, Hi=56, Wi=56, C=64, K=64, Y=3, X=3)
        return (
            WgradConvSpec(
                problem=p,
                tile_m=64,
                tile_n=64,
                tile_k=64,
                warp_m=2,
                warp_n=2,
                warp_tile_m=32,
                warp_tile_n=32,
                warp_tile_k=16,
                pipeline="basic",
                epilogue="default",
                data=ConvDataSpec(dtype_d="fp32"),
                split_k=4,
            ),
            "gfx950",
        )

    if idx == 15:
        # Two-stage deterministic: workspace-store epilogue instead of atomic-add.
        # split_k=4, two_stage=True, fp16 output, gfx950.
        p = ConvProblem(N=8, Hi=56, Wi=56, C=64, K=64, Y=3, X=3)
        return (
            WgradConvSpec(
                problem=p,
                tile_m=64,
                tile_n=64,
                tile_k=64,
                warp_m=2,
                warp_n=2,
                warp_tile_m=32,
                warp_tile_n=32,
                warp_tile_k=16,
                pipeline="mem",
                epilogue="default",
                split_k=4,
                two_stage=True,
            ),
            "gfx950",
        )

    if idx == 16:
        # Two-stage deterministic, gfx942 (16x16x16 MFMA only).
        p = ConvProblem(N=8, Hi=56, Wi=56, C=64, K=64, Y=3, X=3)
        return (
            WgradConvSpec(
                problem=p,
                tile_m=64,
                tile_n=64,
                tile_k=64,
                warp_m=2,
                warp_n=2,
                warp_tile_m=16,
                warp_tile_n=16,
                warp_tile_k=16,
                pipeline="mem",
                epilogue="default",
                split_k=4,
                two_stage=True,
            ),
            "gfx942",
        )

    if idx == 17:
        # gfx1250 wave32 WMMA 16x16x32 K-outer: the transpose read lowers to
        # ds_load_tr16_b128 (8 per lane), so a 16-element fragment is two reads.
        # dtype_d=fp32 because WMMA wgrad supports only the 'default' epilogue,
        # which rejects 16-bit dW.
        from kernels.common._conv_implicit_gemm_common import ConvDataSpec

        p = ConvProblem(N=8, Hi=56, Wi=56, C=64, K=64, Y=3, X=3, pH=1, pW=1)
        return (
            WgradConvSpec(
                problem=p,
                data=ConvDataSpec(dtype_a="fp16", dtype_b="fp16", dtype_d="fp32"),
                tile_m=32,
                tile_n=32,
                tile_k=32,
                warp_m=1,
                warp_n=1,
                warp_tile_m=16,
                warp_tile_n=16,
                warp_tile_k=32,
                wave_size=32,
                pipeline="mem",
                epilogue="default",
                lds_k_outer=True,
            ),
            "gfx1250",
        )

    # ----------------------------------------------------------------
    # Negative cases: these specs must be REJECTED by the validator.
    # The harness (run_emit) expects a ValueError / SystemExit when
    # the config index is >= 100 and "expect_fail=True" is set.
    # ----------------------------------------------------------------
    if idx == 100:
        # Odd C with fp16 split-K -- must raise (packed atomic OOB).
        p = ConvProblem(N=8, Hi=56, Wi=56, C=3, K=64, Y=2, X=2)
        return (
            WgradConvSpec(
                problem=p,
                tile_m=64,
                tile_n=64,
                tile_k=64,
                warp_m=2,
                warp_n=2,
                warp_tile_m=32,
                warp_tile_n=32,
                warp_tile_k=16,
                pipeline="mem",
                epilogue="default",
                split_k=4,
            ),
            "gfx950",
        )
    if idx == 102:
        # split_k > 1 on RDNA gfx1151 -- must raise (CDNA-only).
        p = ConvProblem(N=8, Hi=56, Wi=56, C=64, K=64, Y=3, X=3)
        return (
            WgradConvSpec(
                problem=p,
                tile_m=64,
                tile_n=64,
                tile_k=64,
                warp_m=2,
                warp_n=2,
                warp_tile_m=16,
                warp_tile_n=16,
                warp_tile_k=16,
                wave_size=32,
                pipeline="mem",
                epilogue="default",
                split_k=4,
            ),
            "gfx1151",
        )
    if idx == 103:
        # two_stage=True with split_k=1 -- must raise ValueError.
        p = ConvProblem(N=8, Hi=56, Wi=56, C=64, K=64, Y=3, X=3)
        return (
            WgradConvSpec(
                problem=p,
                tile_m=64,
                tile_n=64,
                tile_k=64,
                warp_m=2,
                warp_n=2,
                warp_tile_m=32,
                warp_tile_n=32,
                warp_tile_k=16,
                pipeline="mem",
                epilogue="default",
                split_k=1,
                two_stage=True,
            ),
            "gfx950",
        )
    raise SystemExit(f"unknown config index {idx}")


def main() -> int:
    return run_emit(
        _spec,
        build_implicit_gemm_conv_wgrad,
        usage="usage: conv_implicit_gemm_wgrad_emit.py <config_index>\n",
    )


if __name__ == "__main__":
    raise SystemExit(main())

File: dnn-providers/hip-kernel-provider/rocke/library/tests/parity/conv_implicit_gemm_wgrad_emit.c
/* Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
 * SPDX-License-Identifier: MIT
 *
 * tests/parity/conv_implicit_gemm_wgrad_emit.c -- C-side emitter for the
 * implicit-GEMM backward-weight convolution parity harness.  Selects one of N
 * sampled spec configs by argv[1] (the config index), builds the
 * rocke_implicit_gemm_conv_wgrad_spec_t identically to the Python emitter
 * conv_implicit_gemm_wgrad_emit.py, builds the kernel via
 * rocke_build_implicit_gemm_conv_wgrad_new and lowers via
 * rocke_lower_kernel_to_llvm (per-config arch, flavor AUTO), printing the .ll
 * to stdout so the two outputs can be byte-compared.
 *
 * Config index table (must stay in sync with the Python emitter):
 *   0  N8H56W56C64_K64Y3X3, t64x64x64, w2x2, a32x32x16, mem/default,      gfx950
 *   1  N8H56W56C64_K64Y3X3, t64x64x64, w2x2, a32x32x16, mem/cshuffle,     gfx950
 *   2  N8H56W56C64_K64Y3X3, t64x64x64, w2x2, a32x32x16, mem/default,      gfx950, split_k=4 fp16
 *   3  N8H56W56C64_K64Y1X1, t64x64x64, w2x2, a32x32x16, mem/default,      gfx950
 *   4  N8H56W56C64_K64Y3X3, t128x128x64, w2x2, a32x32x16, compv4/default, gfx950
 *   5  N8H56W56C64_K64Y3X3, t64x64x64, w2x2, a16x16x16, mem/default,      gfx1151 (WMMA w32)
 *   6  N8H56W56C64_K64Y3X3, t64x64x64, w2x2, a16x16x16, mem/default,      gfx1201 (WMMA w32)
 *   7  N8H56W56C64_K64Y3X3, t64x64x64, w2x2, a32x32x16, mem/default,      gfx950, split_k=4 fp32
 *   8  3-D conv N4Di14H14W14C32_K32Z3Y3X3, t64x64x64, w2x2, a32x32x16, mem/default, gfx950
 *   9  N8H56W56C64_K64Y3X3, t64x64x64, w2x2, a32x32x16, mem/default,      gfx950, split_k=4 bf16
 *  10  N8H56W56C64_K64Y3X3, t64x64x64, w2x2, a32x32x16, mem/default,      gfx950, chiplet_swizzle
 *  11  N8H56W56C64_K64Y3X3, t64x64x64, w2x2, a32x32x16, mem/cshuffle,     gfx950, lds_k_outer
 *  12  N8H56W56C64_K64Y3X3, t64x64x64, w2x2, a32x32x16, mem/default,      gfx950, lds_k_outer + async_dma, split_k=4 fp32
 *  13  N8H56W56C64_K64Y3X3, t64x64x16, w2x2, a16x16x16, mem/cshuffle,     gfx950, lds_k_outer (4 operand elems/lane)
 *  14  N8H56W56C64_K64Y3X3, t64x64x64, w2x2, a32x32x16, basic/default,    gfx950, split_k=4 fp32
 *  15  N8H56W56C64_K64Y3X3, t64x64x64, w2x2, a32x32x16, mem/default,      gfx950, split_k=4 two_stage fp16
 *  16  N8H56W56C64_K64Y3X3, t64x64x64, w2x2, a16x16x16, mem/default,      gfx942, split_k=4 two_stage fp16
 *  17  N8H56W56C64_K64Y3X3 pad1, t32x32x32, w1x1, a16x16x32, mem/default, gfx1250 (WMMA w32), lds_k_outer fp32 out
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "rocke/instance_conv_implicit_gemm_wgrad.h"
#include "rocke/ir.h"
#include "rocke/ir_serialize.h"
#include "rocke/lower_llvm.h"
#include "rocke/verify.h"

/* Fill the config for index `idx`. Returns 0 on success, -1 if unknown.
 * On success sets *spec and *arch. */
static int make_cfg(int idx, rocke_implicit_gemm_conv_wgrad_spec_t* spec, const char** arch)
{
    *spec = rocke_implicit_gemm_conv_wgrad_spec_default();
    spec->tile_m = 64;
    spec->tile_n = 64;
    spec->tile_k = 64;
    spec->warp_m = 2;
    spec->warp_n = 2;
    spec->warp_tile_m = 32;
    spec->warp_tile_n = 32;
    spec->warp_tile_k = 16;
    spec->pipeline = "mem";
    spec->epilogue = "default";

    switch(idx)
    {
    case 0:
        spec->problem = rocke_conv_problem_default(8, 56, 56, 64, 64, 3, 3);
        *arch = "gfx950";
        return 0;
    case 1:
        spec->problem = rocke_conv_problem_default(8, 56, 56, 64, 64, 3, 3);
        spec->epilogue = "cshuffle";
        *arch = "gfx950";
        return 0;
    case 2:
        /* split_k=4, dtype_d="fp16" (default). */
        spec->problem = rocke_conv_problem_default(8, 56, 56, 64, 64, 3, 3);
        spec->split_k = 4;
        *arch = "gfx950";
        return 0;
    case 3:
        spec->problem = rocke_conv_problem_default(8, 56, 56, 64, 64, 1, 1);
        *arch = "gfx950";
        return 0;
    case 4:
        spec->problem = rocke_conv_problem_default(8, 56, 56, 64, 64, 3, 3);
        spec->tile_m = 128;
        spec->tile_n = 128;
        spec->tile_k = 64;
        spec->pipeline = "compv4";
        *arch = "gfx950";
        return 0;
    case 5:
        /* WMMA wave32, gfx1151. */
        spec->problem = rocke_conv_problem_default(8, 56, 56, 64, 64, 3, 3);
        spec->warp_tile_m = 16;
        spec->warp_tile_n = 16;
        spec->warp_tile_k = 16;
        spec->wave_size = 32;
        *arch = "gfx1151";
        return 0;
    case 6:
        /* WMMA wave32, gfx1201. */
        spec->problem = rocke_conv_problem_default(8, 56, 56, 64, 64, 3, 3);
        spec->warp_tile_m = 16;
        spec->warp_tile_n = 16;
        spec->warp_tile_k = 16;
        spec->wave_size = 32;
        *arch = "gfx1201";
        return 0;
    case 7:
        /* split_k=4, dtype_d="fp32". */
        spec->problem = rocke_conv_problem_default(8, 56, 56, 64, 64, 3, 3);
        spec->dtype_d = "fp32";
        spec->split_k = 4;
        *arch = "gfx950";
        return 0;
    case 8:
        /* 3-D convolution: N4 Di14 Hi14 Wi14 C32 K32 Z3 Y3 X3, s=1, p=1, d=1. */
        spec->problem
            = rocke_conv_problem_make_3d(4, 14, 14, 14, 32, 32, 3, 3, 3, 1, 1, 1, 1, 1, 1, 1, 1, 1);
        spec->warp_tile_m = 32;
        spec->warp_tile_n = 32;
        spec->warp_tile_k = 16;
        *arch = "gfx950";
        return 0;
    case 9:
        /* split_k=4, dtype_d="bf16" (packed bf16 atomic path). */
        spec->problem = rocke_conv_problem_default(8, 56, 56, 64, 64, 3, 3);
        spec->dtype_d = "bf16";
        spec->split_k = 4;
        *arch = "gfx950";
        return 0;
    case 10:
        /* chiplet_swizzle enabled. */
        spec->problem = rocke_conv_problem_default(8, 56, 56, 64, 64, 3, 3);
        spec->chiplet_swizzle = true;
        *arch = "gfx950";
        return 0;
    case 11:
        /* K-outer LDS tile + gfx950 ds_read_b64_tr_b16 transpose reads. */
        spec->problem = rocke_conv_problem_default(8, 56, 56, 64, 64, 3, 3);
        spec->epilogue = "cshuffle";
        spec->lds_k_outer = true;
        *arch = "gfx950";
        return 0;
    case 12:
        /* K-outer + direct load (async_dma). */
        spec->problem = rocke_conv_problem_default(8, 56, 56, 64, 64, 3, 3);
        spec->dtype_d = "fp32";
        spec->lds_k_outer = true;
        spec->async_dma = true;
        spec->split_k = 4;
        *arch = "gfx950";
        return 0;
    case 13:
        /* K-outer with the 16x16x16 atom: 4 MFMA operand elements per lane
         * rather than 8, so the transpose-read k-stride between lane groups
         * is 4. Cases 11 and 12 both use 32x32x16 (8 per lane). */
        spec->problem = rocke_conv_problem_default(8, 56, 56, 64, 64, 3, 3);
        spec->tile_k = 16;
        spec->warp_tile_m = 16;
        spec->warp_tile_n = 16;
        spec->warp_tile_k = 16;
        spec->epilogue = "cshuffle";
        spec->lds_k_outer = true;
        *arch = "gfx950";
        return 0;
    case 14:
        /* pipeline="basic": unrolled global-read/compute overlap loop. Needs a
         * compile-time trip count, hence the fixed split_k. */
        spec->problem = rocke_conv_problem_default(8, 56, 56, 64, 64, 3, 3);
        spec->pipeline = "basic";
        spec->dtype_d = "fp32";
        spec->split_k = 4;
        *arch = "gfx950";
        return 0;
    case 15:
        /* Two-stage deterministic: workspace-store epilogue, split_k=4, fp16, gfx950. */
        spec->problem = rocke_conv_problem_default(8, 56, 56, 64, 64, 3, 3);
        spec->split_k = 4;
        spec->two_stage = true;
        *arch = "gfx950";
        return 0;
    case 16:
        /* Two-stage deterministic: workspace-store epilogue, split_k=4, fp16, gfx942
         * (16x16x16 MFMA only). */
        spec->problem = rocke_conv_problem_default(8, 56, 56, 64, 64, 3, 3);
        spec->warp_tile_m = 16;
        spec->warp_tile_n = 16;
        spec->warp_tile_k = 16;
        spec->split_k = 4;
        spec->two_stage = true;
        *arch = "gfx942";
        return 0;
    case 17:
        /* gfx1250 wave32 WMMA 16x16x32 K-outer. The transpose read lowers to
         * ds_load_tr16_b128 (8 per lane), so a 16-element fragment is two
         * reads. dtype_d=fp32 because WMMA wgrad supports only the 'default'
         * epilogue, which rejects 16-bit dW. */
        spec->problem = rocke_conv_problem_default(8, 56, 56, 64, 64, 3, 3);
        spec->problem.pH = 1;
        spec->problem.pW = 1;
        spec->dtype_a = "fp16";
        spec->dtype_b = "fp16";
        spec->dtype_d = "fp32";
        spec->tile_m = 32;
        spec->tile_n = 32;
        spec->tile_k = 32;
        spec->warp_m = 1;
        spec->warp_n = 1;
        spec->warp_tile_m = 16;
        spec->warp_tile_n = 16;
        spec->warp_tile_k = 32;
        spec->wave_size = 32;
        spec->pipeline = "mem";
        spec->epilogue = "default";
        spec->lds_k_outer = true;
        *arch = "gfx1250";
        return 0;
    default:
        return -1;
    }
}

int main(int argc, char** argv)
{
    if(argc < 2)
    {
        fprintf(stderr, "usage: %s <config_index> [ll|ir|verify]\n", argv[0]);
        return 2;
    }
    int idx = atoi(argv[1]);
    const char* mode = (argc > 2) ? argv[2] : "ll";

    rocke_implicit_gemm_conv_wgrad_spec_t spec;
    const char* arch = "gfx950";
    if(make_cfg(idx, &spec, &arch) != 0)
    {
        fprintf(stderr, "unknown config index %d\n", idx);
        return 2;
    }

    rocke_ir_builder_t b;
    rocke_kernel_def_t* kernel = rocke_build_implicit_gemm_conv_wgrad_new(&b, &spec, arch);
    if(kernel == NULL)
    {
        const char* m = rocke_ir_builder_error(&b);
        fprintf(stderr, "build failed: %s\n", m ? m : "(no message)");
        rocke_ir_builder_free(&b);
        return 1;
    }

    int ret = 0;
    if(strcmp(mode, "ll") == 0)
    {
        char* llvm_text = NULL;
        rocke_status_t st
            = rocke_lower_kernel_to_llvm(kernel, ROCKE_LLVM_FLAVOR_AUTO, arch, &llvm_text);
        if(st != ROCKE_OK || !llvm_text)
        {
            fprintf(stderr, "lower failed: status=%d\n", (int)st);
            rocke_ir_builder_free(&b);
            return 1;
        }
        fputs(llvm_text, stdout);
        free(llvm_text);
    }
    else if(strcmp(mode, "ir") == 0)
    {
        char* t = NULL;
        rocke_status_t st = rocke_ir_serialize(kernel, &t);
        if(st != ROCKE_OK || !t)
        {
            fprintf(stderr, "ir_serialize failed: status=%d\n", (int)st);
            rocke_ir_builder_free(&b);
            return 1;
        }
        fputs(t, stdout);
        free(t);
    }
    else if(strcmp(mode, "verify") == 0)
    {
        rocke_diag_t* d = NULL;
        size_t n = 0;
        rocke_verify(kernel, &d, &n);
        for(size_t i = 0; i < n; i++)
        {
            char* s = rocke_diag_to_string(&d[i]);
            if(s)
            {
                puts(s);
                free(s);
            }
        }
        rocke_diags_free(d, n);
    }
    else
    {
        fprintf(stderr, "unknown mode %s\n", mode);
        rocke_ir_builder_free(&b);
        return 2;
    }
    rocke_ir_builder_free(&b);
    return ret;
}
